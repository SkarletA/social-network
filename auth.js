/* eslint-disable import/no-cycle */
/* eslint-disable no-unused-vars */
/* eslint import/no-unresolved: */
import {
  getAuth,
  createUserWithEmailAndPassword,
  GoogleAuthProvider,
  GithubAuthProvider,
  signInWithPopup,
  signInWithEmailAndPassword,
  updateProfile,
  onAuthStateChanged,
  signOut,
} from './firebase.js';
import { onNavigate } from './routes/app.js';
import initApp from './initApp.js';

// Iniciar Firebase
initApp();
const auth = getAuth();
const provider = new GoogleAuthProvider();
const providerGit = new GithubAuthProvider();
auth.languageCode = 'es';

export async function loginGoogle() {
  let results;
  await signInWithPopup(auth, provider)
    .then((result) => {
      results = result;
    })
    .catch((error) => {
      const errorMsg = error.message;
      results = false;
    });
  return results;
}

export async function loginGithub() {
  let results;
  await signInWithPopup(auth, providerGit)
    .then((result) => {
      results = result;
    })
    .catch((error) => {
      const errorMsg = error.message;
      console.log(errorMsg);
      results = false;
    });
  return results;
}

export async function registerUser(email, password) {
  let user;
  await createUserWithEmailAndPassword(auth, email, password)
    .then((userCredential) => {
      user = userCredential.user;
    })
    .catch((error) => {
      const errorCode = error.code;
      const errorMsg = error.message;
    });
  return user;
}

export async function updateUsers(name) {
  await updateProfile(auth.currentUser, {
    displayName: name,
  });
}

export async function loginUser(email, password) {
  let user;
  await signInWithEmailAndPassword(auth, email, password)
    .then((userCredential) => {
      user = userCredential.user;
    })
    .catch((error) => {
      const errorCode = error.code;
      const errorMsg = error.message;
      user = null;
    });
  return user;
}

export function loginUserProfile() {
  const user = auth.currentUser;
  let userValues;
  if (user !== null) {
    userValues = {
      displayName: user.displayName,
      email: user.email,
      photoURL: user.photoURL,
      emailVerified: user.emailVerified,
      uid: user.uid,
    };
  }
  return userValues;
}

export async function logOut() {
  let resultLogOut;
  await signOut(auth)
    .then(() => {
      resultLogOut = true;
    })

    .catch(() => {
      resultLogOut = false;
    });
  return resultLogOut;
}

export const activeSession = () => {
  onAuthStateChanged(auth, (user) => {
    if (user) {
      if (window.location.origin) {
        onNavigate('/home');
      }
    } else {
      onNavigate('/');
    }
  });
};
