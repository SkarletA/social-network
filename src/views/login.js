/* eslint-disable import/no-cycle */
import {
  loginUser, loginUserProfile, loginGoogle, loginGithub,
} from '../auth.js';
import { onNavigate } from '../routes/app.js';
import { register } from '../components/register.js';

export default function login() {
  const expEmail = /^\w+([.+-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,4})+$/;
  const expPassword = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[$@$!%*?&])([A-Za-z\d$@$!%*?&]|[^ ]){8,15}$/;
  const sectionLogin = document.createElement('section');
  sectionLogin.classList.add('login');
  // parrafo de bienvenido
  const pWelcome = document.createElement('p');
  pWelcome.classList.add('welcome');
  pWelcome.innerText = 'Bienvenido';
  // titulo principal del proyecto
  const titleProject = document.createElement('h1');
  titleProject.innerText = 'Mapask';
  // parrafo del eslogan
  const pSlogan = document.createElement('p');
  pSlogan.innerText = 'Conecta con personas que se apasionan por la programación';
  // contenedor de imagen principal
  const containerImageLogin = document.createElement('picture');
  containerImageLogin.classList.add('image-login');
  const imageLogin = document.createElement('img');
  imageLogin.src = 'https://svgshare.com/i/fmB.svg';
  imageLogin.title = 'img-login';
  // contenedor de logueo
  const containerLogin = document.createElement('section');
  containerLogin.classList.add('container-login');
  // titulo secondario
  const titleProjectSecond = document.createElement('h2');
  titleProjectSecond.innerText = 'Mapask';
  // Formulario de logueo
  const formLogin = document.createElement('form');
  formLogin.classList.add('form-session');
  formLogin.onsubmit = 'return false';
  // input de email
  const formInputEmail = document.createElement('input');
  formInputEmail.classList.add('input');
  formInputEmail.classList.add('input-email');
  formInputEmail.id = 'inputEmail';
  formInputEmail.type = 'text';
  formInputEmail.placeholder = 'Email: ';
  formInputEmail.required = 'required';
  // contenedor para la alerta de email
  const containerAlertlEmail = document.createElement('div');
  containerAlertlEmail.classList.add('container-email');
  containerAlertlEmail.id = 'containerEmail';
  // input de password
  const formInputPassword = document.createElement('input');
  formInputPassword.classList.add('input');
  formInputPassword.classList.add('input-password');
  formInputPassword.id = 'inputPassword';
  formInputPassword.type = 'password';
  formInputPassword.placeholder = 'Contraseña: ';
  // contenedor para la alerta de password
  const containerAlertlPassword = document.createElement('div');
  containerAlertlPassword.classList.add('container-password');
  containerAlertlPassword.id = 'containerPassword';
  // boton de iniciar sesion
  const formButton = document.createElement('button');
  formButton.classList.add('button');
  formButton.classList.add('btn-login');
  formButton.id = 'btnLogin';
  formButton.type = 'password';
  formButton.innerText = 'Iniciar Sesion ';
  // contenedor de link de registro
  const divRegistration = document.createElement('div');
  divRegistration.classList.add('registration');
  const spanMessage = document.createElement('span');
  spanMessage.innerText = '¿No estas en Mapask?';
  const linkRegistration = document.createElement('a');
  linkRegistration.classList.add('link-registration');
  linkRegistration.id = 'linkRegistration';
  linkRegistration.innerText = 'Registrate';

  // contenerdor de botones auth
  const btnsAuthContainer = document.createElement('div');
  btnsAuthContainer.classList.add('btns-auth-container');
  // contenedor de boton de google
  const buttonGoogle = document.createElement('button');
  buttonGoogle.classList.add('btn-google');
  buttonGoogle.id = 'btnGoogle';
  // buttonGoogle.innerText = 'GOOGLE';
  const spanIconGoogle = document.createElement('img');
  spanIconGoogle.classList.add('iconGoogle');
  spanIconGoogle.src = 'https://svgshare.com/i/fnL.svg';

  // contenedor de boton de Github
  const buttonGit = document.createElement('button');
  buttonGit.classList.add('btn-google');
  buttonGit.id = 'btnGit';
  const spanIconGit = document.createElement('img');
  spanIconGit.classList.add('iconGit');
  spanIconGit.src = 'https://svgshare.com/i/fmC.svg';
  buttonGit.appendChild(spanIconGit);
  btnsAuthContainer.appendChild(buttonGoogle);
  btnsAuthContainer.appendChild(buttonGit);

  // Contenedor alerta de Google
  const alertGoogle = document.createElement('div');

  containerImageLogin.appendChild(imageLogin);

  formLogin.appendChild(formInputEmail);
  formLogin.appendChild(containerAlertlEmail);
  formLogin.appendChild(formInputPassword);
  formLogin.appendChild(containerAlertlPassword);
  formLogin.appendChild(formButton);

  divRegistration.appendChild(spanMessage);
  divRegistration.appendChild(linkRegistration);
  buttonGoogle.appendChild(spanIconGoogle);
  containerLogin.appendChild(titleProjectSecond);
  containerLogin.appendChild(formLogin);
  containerLogin.appendChild(divRegistration);

  containerLogin.appendChild(btnsAuthContainer);
  containerLogin.appendChild(alertGoogle);

  sectionLogin.appendChild(pWelcome);
  sectionLogin.appendChild(titleProject);
  sectionLogin.appendChild(pSlogan);
  sectionLogin.appendChild(containerImageLogin);
  sectionLogin.appendChild(containerLogin);

  formButton.addEventListener('click', async (e) => {
    e.preventDefault();
    const email = formInputEmail.value;
    const password = formInputPassword.value;

    if (expEmail.test(email) && expPassword.test(password)) {
      const user = await loginUser(email, password);
      if (!user) {
        containerAlertlPassword.innerHTML = '<span class="red"> Usuario no registrado </span>';
      } else {
        containerAlertlPassword.innerHTML = '';
        onNavigate('/home');
        const userId = (loginUserProfile()).uid;
        localStorage.setItem('userId', userId);
      }
    } else {
      containerAlertlPassword.innerHTML = '<span class="red"> Correo o constraseña inválido </span>';
    }
  });

  buttonGoogle.addEventListener('click', async (e) => {
    e.preventDefault();
    const userGoogle = await loginGoogle();
    if (!userGoogle) {
      alertGoogle.innerHTML = '<span class="red"> Error al iniciar sesión </span>';
    } else {
      alertGoogle.innerHTML = '';
      onNavigate('/home');
    }
  });

  buttonGit.addEventListener('click', async (e) => {
    e.preventDefault();
    const userGit = await loginGithub();
    if (!userGit) {
      alertGoogle.innerHTML = '<span class="red"> Error al iniciar sesión </span>';
    } else {
      alertGoogle.innerHTML = '';
      onNavigate('/home');
      console.log('me logee');
    }
  });

  linkRegistration.addEventListener('click', () => {
    sectionLogin.appendChild(register());
  });

  return sectionLogin;
}
