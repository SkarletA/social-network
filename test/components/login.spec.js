/**
 * @jest-environment jsdom
 */
import { loginUser, loginGoogle } from '../../src/auth.js';
import login from '../../src/views/login.js';

jest.mock('../../src/firebase.js');

describe('login', () => {
  const result = login();
  it('login deberia ser una función', () => {
    expect(typeof login).toBe('function');
  });

  it('deberia dar un error si las contraseñas no coinciden', () => {
    const email = result.querySelector('#inputEmail');
    const password = result.querySelector('#inputPassword');

    email.value = 'gsmaggie001gmail.com';
    password.value = 't123@mapi';

    const btn = result.querySelector('#btnLogin');
    btn.dispatchEvent(new Event('click'));

    const alert = result.querySelector('#containerPassword');
    alert.innerHTML = '<span class="red"> Correo o constraseña inválido </span>';
    expect(alert.textContent).toBe(' Correo o constraseña inválido ');
  });

  document.body.innerHTML = '<div id="root"></div>';
  test('render de login', () => {
    const root = document.getElementById('root');
    root.appendChild(result);
    expect(root.innerHTML).toMatchSnapshot();
  });

  const email = result.querySelector('#inputEmail');
  const password = result.querySelector('#inputPassword');
  test('Te devuelve el id del usuario si esta registrado', async () => {
    email.value = 'gsmaggie001@gmail.com';
    password.value = 'T123@maps';
    const user = await loginUser(email, password);
    expect(user).toEqual({ uid: 'za123' });
  });

  test('Deberia dar error al no estar registrado el usuario', async () => {
    email.value = 'gsmaggie001@gmail.com';
    password.value = 'T123@maps';
    try {
      await loginUser(email, password);
    } catch (e) {
      expect(e).toMatch(' Usuario no registrado ');
    }
  });

  test('Debería devolver el id del usuario', async () => {
    const userGoogle = await loginGoogle();
    expect(userGoogle).toEqual({ uid: 'za123' });
  });

  test('Debería dar un error al no temrinar el inicio de sesión', async () => {
    try {
      await loginGoogle();
    } catch (e) {
      expect(e).toMatch(' Error al iniciar sesión ');
    }
  });
});
