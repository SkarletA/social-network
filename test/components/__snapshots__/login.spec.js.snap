// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login render de login 1`] = `"<section class=\\"login\\"><p class=\\"welcome\\"></p><h1></h1><p></p><picture class=\\"image-login\\"><img src=\\"https://svgshare.com/i/ept.svg\\" title=\\"img-login\\"></picture><section class=\\"container-login\\"><h2></h2><form class=\\"form-session\\"><input class=\\"input input-email\\" id=\\"inputEmail\\" type=\\"text\\" placeholder=\\"Email: \\" required=\\"\\"><div class=\\"container-email\\" id=\\"containerEmail\\"></div><input class=\\"input input-password\\" id=\\"inputPassword\\" type=\\"password\\" placeholder=\\"Contraseña: \\"><div class=\\"container-password\\" id=\\"containerPassword\\"><span class=\\"red\\"> Correo o constraseña inválido </span></div><button class=\\"button btn-login\\" id=\\"btnLogin\\" type=\\"submit\\"></button></form><div class=\\"registration\\"><span></span><a class=\\"link-registration\\" id=\\"linkRegistration\\"></a></div><button class=\\"btn-google\\" id=\\"btnGoogle\\"><span class=\\"iconify\\" data-icon=\\"akar-icons:google-contained-fill\\"></span></button><div></div></section></section>"`;
